Fixed 4616 : Option for editing depictions

Ayan-10 opened the pull request:
**Description (required)**

Made option for editing depictions.

Fixes #4616 

What changes did you make and why?

In the current app depicts can be added only while uploading, it's very necessary for a user to edit depictions. 

1. I reused the depiction selection UI which is being used while uploading for letting user select the depicts
2. Then I compare it with old depicts and send only new depicts for posting to the server
3. A POST API call will post the new depicts in the commons server

**Tests performed (required)**

Tested lastest master betaDebug and ProdDebug on Pixel 3 with API level 30.

# TODO

All tested with latest version after resetting data

- [x] Adding two depictions, actually only adds one.
- [x]     Pressing "Back" goes back one level too far
- [ ]     Pressing "Back" is often very slow. If you wait for a few minutes on the UI, pressing the "Back" button in the depictions editor sometimes does nothing at first, the UI freezes.
- [x]     The Nearby top banner sometimes appears, for instance if you open the dialog fast enough after starting the app.
- [x]     Open the UI for a picture, close it, open for another picture: The depictions of the previous picture are still visible for a few seconds. They should not appear.
- [x]     The UI does not take the whole screen: A bit of the media details is still visible at the bottom.
- [x]     After adding a new depiction and saving, then opening the edition dialog again, the checkbox is not checked anymore.
- [x]     When editing several pictures in a row, the edits are erroneous, for instance a picture for which I checked and saved "street" got actually edited (when looking at the Commons website) as "beach", a depiction I had used for a different picture before.
- [x] Depictions added via edit take longer to load in Media Details than the depictions that had been added in the Upload Wizard. For instance if I added depiction1/2/3 with Upload Wizard and depiction4/5/6 with the edition UI, then depiction1/2/3 appear immediately but depiction4/5/6 take a second or two to load. There should be zero difference.
- [x] Can not _remove_ a depiction.
- [ ] When I reach license then go back to depictions, sometimes the depictions I had selected are not selected anymore.

nicolas-raoul reviewed (commented):
Thanks, that is extremely helpful!
I noticed a few issues:
- After saving, depiction does not appear immediately in the media details (above the pen).
- Pressing the "Back" button in the depictions editor sometimes does nothing at first, the UI freezes, then all of the backs are done very quickly. I observed something similar in the categories editor pull request.
- The nearby banner appears at the bottom, if you open the dialog fast enough after starting the app.

Ayan-10 commented:
@nicolas-raoul I think "Depictions not updating instantly" issue is resolved now.

nicolas-raoul commented:
Would you mind solving the conflict? Sorry for the delay in reviewing, and thanks for your patience!

Ayan-10 commented:
@nicolas-raoul I resolved the conflicts and also could you please review #4515 once? the requested changes are made a long ago.

nicolas-raoul reviewed (changes requested):
I noticed another issue: On a picture that already has a depiction, the Edit Depicts dialog does not show that depiction. If I search for it, it is shown as checked.
Desired behavior: When opening Edit Depicts, it should show all current depictions (with the checkbox checked).
Thanks a lot!

nicolas-raoul reviewed (changes requested):
Also, pressing Back should bring back to the media details.
Currently, pressing back brings me to the list of media (one level too far).

Ayan-10 commented:
> I noticed another issue: On a picture that already has a depiction, the Edit Depicts dialog does not show that depiction. If I search for it, it is shown as checked. Desired behavior: When opening Edit Depicts, it should show all current depictions (with the checkbox checked). Thanks a lot!

To show all existing depicts at the top we have to make a call to get DepictedItems because only Depicted IDs are available as existing depicts. In this PR I just compared the existing IDs with the search result and changed their checked states accordingly.

nicolas-raoul commented:
So you would have to call the API (with the depicted IDs) to get the information required for displaying them, right?

That sounds doable, and I think it is important to have a consistent behavior with the Upload Wizard :-)

nicolas-raoul commented:
Great, the UI now shows already existing depictions! :-)

Remaining bugs:

- Pressing "Back" goes back one level too far
- If you wait for a few minutes on the UI, pressing the "Back" button in the depictions editor sometimes does nothing at first, the UI freezes.
- The Nearby top banner sometimes appears, for instance if you open the dialog fast enough after starting the app.
- Open the UI for a picture, close it, open for another picture: The depictions of the previous picture are still visible for a few seconds. They should not appear.
- The UI does not take the whole screen: A bit of the media details is still visible at the bottom.

Ayan-10 commented:
> * Open the UI for a picture, close it, open for another picture: The depictions of the previous picture are still visible for a few seconds. They should not appear.

@nicolas-raoul Currently the depictions of the previous picture are appearing as checked. I will remove the checked state of previous depictions. Is this ok with you or the previous ones should not appear at all?

nicolas-raoul commented:
> I will remove the checked state of previous depictions

Sounds good! :-)

nicolas-raoul reviewed (changes requested):
Depiction search in the Upload Wizard does not work anymore:
![Screenshot_20211231-122103_Commons](https://user-images.githubusercontent.com/99590/147801291-de7f9bf4-9301-4a31-bdb0-08b8ba738d31.jpg)

nicolas-raoul reviewed (changes requested):
Depictions of the previous picture currently appear for a second (unchecked) then disappear.
This is not satisfying: they should either appear and remain visible as suggestions, or not appear at all (not even briefly)

Ayan-10 commented:
@nicolas-raoul Sorry for the inconvenience. I should have checked the upload wizard. I fixed the upload wizard and previous depicts issue. You can review it.

nicolas-raoul reviewed (changes requested):
Another problem:
1. Add 2 depictions
2. Press "Save"
3. Only one of the depictions is added (both to the server and in the UI).

Please don't take it as an offense, but I think you should test more thoroughly before asking for a review.
I added a "TODO" section at https://github.com/commons-app/apps-android-commons/pull/4725#issue-799945122, would you mind fixing all of the points (you can check the points solved, one by one) then ask for review? Thanks! :-)

Ayan-10 commented:
Sorry, I will keep that in mind.

Ayan-10 commented:
@nicolas-raoul I checked all points of the TODO list one by one and fixed all of them.

1. Only one depiction addition problem is fixed, multiple depictions can be added now.
2. Pressing "Back" goes back to MediaDetailFragment.
3. After making recent changes on the back button, never occurred for me from then.
4. The Nearby top banner is disabled in DepictsFragment.
5. Previous Screen is cleared, Previous depictions are not visible in the recent fragment.
6. The UI is taking the whole screen now.
7. After the latest changes on this PR never occurred for me.
8. Edited 5 pictures. Didn't occur for me after a recent change in logic. You can check those edited files here.
https://commons.wikimedia.org/wiki/Special:Contributions/Ayan184

Tested and fixed some minor bugs also. I hope the PR is ready for review.

Thanks.

Tested prodDebug in Pixel 3 API 30 emulator.

nicolas-raoul reviewed (changes requested):
7 is still happening on first try.
Back is sometimes very slow.

Ayan-10 commented:
https://user-images.githubusercontent.com/71203077/147866784-de4964c1-2980-430d-a5ee-ac139d8a7903.mp4

@nicolas-raoul As you can see in the video 
1. back button pressing is not taking a long time 
2. 7 is not occurring also as you can see I immediately opened the editor after editing and all selected depicts are showing as expected.

I tested in my real device and emulator too. Issues are not occurring.

Don't know why it's occurring for you. Can you please uninstall the app then install this PR version.

> * Depictions added via edit take longer to load in Media Details than the others, even after restarting the app. There should be zero difference.

The last TODO that you added is not a problem of this PR. In master also the depictions will take a little more time to load compared to others because It's making a GET call to the server to get depicts from the available IDs. It's not an issue of this PR.

nicolas-raoul commented:
I updated the TODO list.
You should try to reproduce the bugs I mentioned, you will have to try several ways (for instance editing depictions of your contributions), testing can take time (usually the same amount of time as developing).

Ayan-10 commented:
@nicolas-raoul I got your 9th point. I can able to reproduce 9. Sometimes it occurred only when MediaDetailFragment is opening from own contributions. I fixed the issue. 

But still, I am not able to reproduce 3 and 8 (Ui freeze on back button press, the edits are erroneous). Maybe I need to test more on own contributions to reproduce 3 and 8. As I have very few own contributions it's hard to test on it. Is it possible to give a Commons ID with a lot of own contributions through which I can test the contributions?

Thanks.

nicolas-raoul commented:
I haven't noticed 8 in a while indeed, let's hope it is fixed. :-)

I have a test account with hundreds of uploads, I can send it to you if you let me know of an email address.

Ayan-10 commented:
@nicolas-raoul I have sent you my mail Id in Zulip.

Ayan-10 commented:
> * Depictions do not appear in Airplane mode (no Internet connection). Categories appear correctly even in Airplane mode, on the other hand. Maybe in the example above, depiction4/5/6 should be stored locally at the same place as depiction1/2/3

@nicolas-raoul Do depicts appear without the internet connection in master? If I am not wrong, depicts are not appearing without the internet in master too, as the app is fetching depicts every time from the server. 

Thanks

nicolas-raoul commented:
Indeed you are right about offline! I removed the task. I added another one though.
Thanks!

Ayan-10 commented:
> * Can not _remove_ a depiction.

@nicolas-raoul Actually that was intended. As current edit categories don't have an option to remove a category, I thought the remove option in depiction is not required too. But I will surely implement the remove feature asap.

Thanks.

nicolas-raoul reviewed (changes requested):
Thanks for adding the remove feature, that's great!

I can still reproduce the freeze. Steps:
1. Open Contributions
2. Click the depiction edit button
3. Waiting for 10 minutes doing nothing
4. (screen got locked so I unlocked it)
5. Press back button to go back to MediaDetailFragment. It works so far.
6. Press back button to go back to Explore: It takes about a minute or more to work. 

https://user-images.githubusercontent.com/99590/149118830-b08e3ae2-740f-4042-b7e4-40bbb8a7cc62.mp4

Ayan-10 commented:
@nicolas-raoul I found another issue, if the keypad is open in depicts editor, the back button press is not working properly. I fixed that issue. 

Still can't reproduce the UI freeze issue, tried a lot of times with the exact same steps that you mentioned.
I will try to reproduce it again. But I would like to ask can you observe any other step other than that you mentioned here? and is the issue occurring for you every time?

nicolas-raoul commented:
I tried again, the freeze still happens every single time when waiting for long enough.

nicolas-raoul commented:
Steps showing the slowdown:
1. Use a Wikimedia account with hundreds of uploads
2. Open the app
3. In Contributions, tap the first picture
4. Tap the depiction edit button
5. Wait for 5 minutes doing nothing
6. Press back button to go back to MediaDetailFragment. It works so far.
7. Press back button to go back to Explore: It takes 3 seconds, whereas it should be immediate.

I have used this branch for a few days, and have been able to reproduce this many times. The more I wait at step 5, the most severe the freeze is.

Also, in the Upload Wizard, when I reach license then go back to depictions, sometimes the depictions I had selected are not selected anymore. I will now switch to master for a while to see whether that happens in master too or not.

Thanks!

Ayan-10 commented:
@nicolas-raoul Can you please mention the device and API level on which you are facing those issues? I had used `@RequiresApi(api = VERSION_CODES.O)` annotation in some places. Maybe that can be a reason that I can't reproduce it for a single time after trying so many times following exact same steps as I am testing it on API Level 30.

nicolas-raoul commented:
@Ayan-10 I am reproducing on Android 9.

nicolas-raoul commented:
Would you mind rebasing? Thanks :-)

nicolas-raoul commented:
Unfortunately I am still getting the freeze problem, on both Android 9 and Android 12, with the exact steps mentioned at https://github.com/commons-app/apps-android-commons/pull/4725#issuecomment-1023097726 .

nicolas-raoul commented:
Sorry would you mind rebasing again? Thanks!

nicolas-raoul commented:
I am still getting long freezes unfortunately :-/

nicolas-raoul commented:
Have you tried on the account with hundreds of contributions?

Ayan-10 commented:
@nicolas-raoul Yes, I am testing with the commons user ID that you have given me.

nicolas-raoul commented:
Here is my logcat when I get a short slowdown of a few seconds:
```
02-26 10:28:33.175 13262 13321 D OkHttp  : <-- 200 https://commons.wikimedia.org/w/api.php?format=json&action=wbgetentities&props=labels&languagefallback=1&ids=M113255462%7CM113255474%7CM113255484%7CM113255498%7CM113255525%7CM113255551%7CM113255582%7CM113255613%7CM113255644%7CM113255653 (734ms, unknown-length body)
02-26 10:28:33.357 13262 13461 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allimages&gaisort=timestamp&gaidir=older&prop=imageinfo&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl&gaiuser=Syced&gailimit=10&gaicontinue=20211211092533|Walk_in_Shirokane_3.jpg&continue=gaicontinue||
02-26 10:28:33.360 13262 13461 D CookieManager: Domain:commons.wikimedia.org
[... such lines keep appearing at a fast pace during the whole time I am on the Media Details screen ...]
02-26 10:28:34.090 13262 13461 D OkHttp  : <-- 200 https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allimages&gaisort=timestamp&gaidir=older&prop=imageinfo&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl&gaiuser=Syced&gailimit=10&gaicontinue=20211211092533|Walk_in_Shirokane_3.jpg&continue=gaicontinue|| (733ms, unknown-length body)
02-26 10:28:34.100 13262 13461 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?format=json&action=wbgetentities&props=labels&languagefallback=1&ids=M113167708%7CM113222361%7CM113222397%7CM113251888%7CM113252200%7CM113252948%7CM113255295%7CM113255399%7CM113255440%7CM113255453
02-26 10:28:34.104 13262 13461 D CookieManager: Domain:commons.wikimedia.org
[ I tapped "Back" around here]
02-26 10:28:34.342  5636  5803 I clp-JNI : Add pinfo.  (channel) e66e241 fr.free.nrw.commons/fr.free.nrw.commons.contributions.MainActivity (server) (pkg) AppWindowToken{7251a86 token=Token{c4a1f61 ActivityRecord{b9d12c8 u0 fr.free.nrw.commons/.contributions.MainActivity t33938}}} 13262 10408 (action) 0
02-26 10:28:34.347 13262 13262 D ViewRootImpl@97e33ea[MainActivity]: ViewPostIme key 0
02-26 10:28:34.396  5636  5803 I clp-JNI : Add pinfo.  (channel) e66e241 fr.free.nrw.commons/fr.free.nrw.commons.contributions.MainActivity (server) (pkg) AppWindowToken{7251a86 token=Token{c4a1f61 ActivityRecord{b9d12c8 u0 fr.free.nrw.commons/.contributions.MainActivity t33938}}} 13262 10408 (action) 1
02-26 10:28:34.397 13262 13262 D ViewRootImpl@97e33ea[MainActivity]: ViewPostIme key 1
02-26 10:28:34.851 13262 13461 D OkHttp  : <-- 200 https://commons.wikimedia.org/w/api.php?format=json&action=wbgetentities&props=labels&languagefallback=1&ids=M113167708%7CM113222361%7CM113222397%7CM113251888%7CM113252200%7CM113252948%7CM113255295%7CM113255399%7CM113255440%7CM113255453 (750ms, unknown-length body)
02-26 10:28:38.032 13262 13449 D OkHttp  : --> GET https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park
02-26 10:28:38.032 13262 13357 D OkHttp  : --> GET https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park
02-26 10:28:38.034 13262 13357 D CookieManager: Domain:en.wikipedia.org
02-26 10:28:38.034 13262 13357 D CookieManager: Adding centralauth cookies
02-26 10:28:38.035 13262 13449 D CookieManager: Domain:en.wikipedia.org
02-26 10:28:38.035 13262 13449 D CookieManager: Adding centralauth cookies
02-26 10:28:38.055 13262 14160 V OkHttp  : --> GET https://upload.wikimedia.org/wikipedia/commons/thumb/6/62/Tabata_station.jpg/640px-Tabata_station.jpg
02-26 10:28:38.055 13262 14160 V OkHttp  : Cache-Control: no-store
02-26 10:28:38.055 13262 14160 V OkHttp  : --> END GET
02-26 10:28:38.063 13262 14163 V OkHttp  : --> GET https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/Asukayama_Park_30.jpg/640px-Asukayama_Park_30.jpg
02-26 10:28:38.063 13262 14163 V OkHttp  : Cache-Control: no-store
02-26 10:28:38.063 13262 14163 V OkHttp  : --> END GET
02-26 10:28:38.066 13262 14163 V OkHttp  : <-- HTTP FAILED: java.io.IOException: Canceled
02-26 10:28:38.067 13262 14160 V OkHttp  : <-- HTTP FAILED: java.io.IOException: Canceled
02-26 10:28:38.072 13262 14165 V OkHttp  : --> GET https://upload.wikimedia.org/wikipedia/commons/thumb/6/6e/Asukayama_Park_29.jpg/640px-Asukayama_Park_29.jpg
02-26 10:28:38.073 13262 13331 I OpenGLRenderer: Davey! duration=3668ms; Flags=0, IntendedVsync=53487588813203, Vsync=53487605479869, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=53487607744430, AnimationStart=53487607758738, PerformTraversalsStart=53487613645123, DrawStart=53491218793351, SyncQueued=53491220639082, SyncStart=53491220988082, IssueDrawCommandsStart=53491240942351, SwapBuffers=53491255649735, FrameCompleted=53491257625312, DequeueBufferDuration=980000, QueueBufferDuration=410000, 
02-26 10:28:38.074 13262 14165 V OkHttp  : Cache-Control: no-store
02-26 10:28:38.074 13262 14165 V OkHttp  : --> END GET
02-26 10:28:38.075 13262 14165 V OkHttp  : <-- HTTP FAILED: java.io.IOException: Canceled
02-26 10:28:38.154 13262 13450 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allimages&gaisort=timestamp&gaidir=older&prop=imageinfo&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl&gaiuser=Syced&gailimit=10&gaicontinue=20211210065107|Constructrion_at_Shibuya_station_17.jpg&continue=gaicontinue||
02-26 10:28:38.157 13262 13450 D CookieManager: Domain:commons.wikimedia.org
02-26 10:28:38.173 13262 13463 D OkHttp  : --> GET https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park
02-26 10:28:38.178 13262 13463 D CookieManager: Domain:en.wikipedia.org
02-26 10:28:38.178 13262 13463 D CookieManager: Adding centralauth cookies
02-26 10:28:38.178 13262 13475 D OkHttp  : --> GET https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park
02-26 10:28:38.180 13262 14165 V OkHttp  : --> GET https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/Asukayama_Park_30.jpg/640px-Asukayama_Park_30.jpg
02-26 10:28:38.180 13262 14165 V OkHttp  : Cache-Control: no-store
02-26 10:28:38.181 13262 14165 V OkHttp  : --> END GET
02-26 10:28:38.182 13262 13475 D CookieManager: Domain:en.wikipedia.org
02-26 10:28:38.182 13262 13475 D CookieManager: Adding centralauth cookies
02-26 10:28:38.183 13262 14160 V OkHttp  : --> GET https://upload.wikimedia.org/wikipedia/commons/thumb/6/6e/Asukayama_Park_29.jpg/640px-Asukayama_Park_29.jpg
02-26 10:28:38.183 13262 14160 V OkHttp  : Cache-Control: no-store
02-26 10:28:38.183 13262 14160 V OkHttp  : --> END GET
02-26 10:28:38.190 13262 14163 V OkHttp  : --> GET https://upload.wikimedia.org/wikipedia/commons/thumb/6/62/Tabata_station.jpg/640px-Tabata_station.jpg
02-26 10:28:38.190 13262 14163 V OkHttp  : Cache-Control: no-store
02-26 10:28:38.190 13262 14163 V OkHttp  : --> END GET
02-26 10:28:38.203 13262 14163 V OkHttp  : <-- HTTP FAILED: java.io.IOException: Canceled
02-26 10:28:38.203 13262 14165 V OkHttp  : <-- HTTP FAILED: java.io.IOException: Canceled
02-26 10:28:38.204 13262 14160 V OkHttp  : <-- HTTP FAILED: java.io.IOException: Canceled
02-26 10:28:38.298 13262 13357 D OkHttp  : <-- 200 https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park (265ms, unknown-length body)
02-26 10:28:38.427 13262 13449 D OkHttp  : <-- 200 https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park (394ms, unknown-length body)
02-26 10:28:38.427 13262 13463 D OkHttp  : <-- 200 https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park (254ms, unknown-length body)
02-26 10:28:38.428 13262 13475 D OkHttp  : <-- 200 https://en.wikipedia.org/api/rest_v1/page/media-list/Asukayama_Park (249ms, unknown-length body)
02-26 10:28:38.744 13262 13450 D OkHttp  : <-- 200 https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allimages&gaisort=timestamp&gaidir=older&prop=imageinfo&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl&gaiuser=Syced&gailimit=10&gaicontinue=20211210065107|Constructrion_at_Shibuya_station_17.jpg&continue=gaicontinue|| (589ms, unknown-length body)
02-26 10:28:38.754 13262 13450 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?format=json&action=wbgetentities&props=labels&languagefallback=1&ids=M113218844%7CM113218860%7CM113218866%7CM113218879%7CM113218899%7CM113218913%7CM113218921%7CM113218941%7CM113218949%7CM113218959
02-26 10:28:38.756 13262 13450 D CookieManager: Domain:commons.wikimedia.org
02-26 10:28:39.697 13262 13450 D OkHttp  : <-- 200 https://commons.wikimedia.org/w/api.php?format=json&action=wbgetentities&props=labels&languagefallback=1&ids=M113218844%7CM113218860%7CM113218866%7CM113218879%7CM113218899%7CM113218913%7CM113218921%7CM113218941%7CM113218949%7CM113218959 (942ms, unknown-length body)
02-26 10:28:39.835 13262 13330 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allimages&gaisort=timestamp&gaidir=older&prop=imageinfo&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl&gaiuser=Syced&gailimit=10&gaicontinue=20211210064112|Constructrion_at_Shibuya_station_6.jpg&continue=gaicontinue||
```
The [OpenGLRenderer: Davey](https://stackoverflow.com/questions/57132218/openglrenderer-davey) is suspicious, and the milliseconds duration seems to match. I wonder why this is happening in this branch and not in master.

nicolas-raoul commented:
After waiting for a few minutes on the Media Details screen before tapping "Back", the freeze lasted over a minute before this appeared:
```
02-26 10:49:19.659 13262 13331 I OpenGLRenderer: Davey! duration=162198ms; Flags=0, IntendedVsync=54570643449818, Vsync=54570643449818, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=54570651152079, AnimationStart=54570651159463, PerformTraversalsStart=54570656188963, DrawStart=54732801130463, SyncQueued=54732808677270, SyncStart=54732810726117, IssueDrawCommandsStart=54732834149540, SwapBuffers=54732841741886, FrameCompleted=54732843760655, DequeueBufferDuration=1233000, QueueBufferDuration=492000, 
```
![adb](https://user-images.githubusercontent.com/99590/155824312-a8db26a4-782f-4db8-9550-d5d6a469b109.png)

nicolas-raoul commented:
"_after pressing the cancel button_" -> It happens also after doing no edit and just tapping cancel.

I believe it does not reproduce if there are no pictures downloading in the background.
100 pictures is not a huge number, so probably the 100 pictures were downloaded already when you tried.
Could you please try this?
1. Reset the app's storage (in Settings)
2. Log in
3. Immediately tap the first picture, tap the edit icon for depictions
4. Wait for 60 seconds
5. Tap Cancel
6. Wait for 60 seconds
7. Press "Back", see whether there is a delay

If it does not work, try again (resetting again) with a longer wait at step 6.
Thanks a lot!

Ayan-10 commented:
@nicolas-raoul Sorry for the late response and thanks for the mentioned steps.

I finally reproduced the issue with these steps. But I wonder why it's happening only with this branch. Till exiting the depicts screen everything is normal and as expected. But after clicking the back button on `MediaDetailFragment` screen it's showing some lagging. But all the network calls are the same as master after exiting the depicts screen. Also, I waited longer on step 6 but freezing time doesn't increase it's the same as waiting for 60 sec. As you have mentioned about the 100 pictures fact, it seems the freeze is happening because of that part. Although I didn't change anything on that part I will still try to figure out something from that part to solve the bug. 

Also, today I am not able to reproduce the issue with these steps but I was able to reproduce it some day earlier.

nicolas-raoul commented:
It is a very strange issue indeed.

I suspect that the root cause is that the app does many downloads in the background. I created a different issue about that: #4900 

So, I think it would be fair to merge this PR even with this bug, if other quality conditions are met (in particular unit tests coverage).

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:
```diff
@@ -517,9 +519,28 @@ private void onDeletionPageExists(Boolean deletionPageExists) {
+        final FragmentTransaction transaction = getChildFragmentManager().beginTransaction();
+        transaction.replace(R.id.mediaDetailFrameLayout, depictsFragment);
+        transaction.addToBackStack(null);
+        transaction.commit();
```
unit test?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:
```diff
@@ -1140,73 +1161,73 @@ public void updateCoordinates(final String Latitude, final String Longitude,
+            builder.setNegativeButton(R.string.about_translate_cancel,
+                (dialog, which) -> dialog.dismiss());
+            AlertDialog dialog = builder.create();
+            dialog.show();
```
unit test?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java:
```diff
@@ -275,6 +293,39 @@ public void onDepictItemClicked(DepictedItem depictedItem) {
+     * @return string ex. "Q11023|Q1356"
+     */
+    private String joinIDs(final List<String> depictIDs) {
+        if (depictIDs != null && !depictIDs.isEmpty()) {
```
unit test

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/UploadModel.java:
```diff
@@ -185,11 +191,33 @@ public void deletePicture(final String filePath) {
+                        final List<String> depictsList = new ArrayList<>();
+                        depictsList.add(depictedItem.getId());
+                        depictsList.addAll(media.getDepictionIds());
+                        media.setDepictionIds(depictsList);
```
unit tests?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictEditHelper.kt:
```diff
@@ -0,0 +1,121 @@
+     * @return Observable<Boolean>
+     */
+     private fun addDepiction(media: Media, depicts: List<String>): Observable<Boolean> {
+        Timber.d("thread is depict adding %s", Thread.currentThread().name)
```
"thread is depict adding %s" -> "thread is adding depiction %s"

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictEditHelper.kt:
```diff
@@ -0,0 +1,121 @@
+        context: Context,
+        media: Media,
+        result: Boolean
+    ): Boolean {
```
unit tests?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsFragment.java:
```diff
@@ -133,19 +188,28 @@ public void goToPreviousScreen() {
-            this::goToNextScreen,
-            null
-        );
+        if (media == null) {
```
unit tests?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsPresenter.kt:
```diff
@@ -71,16 +78,37 @@ class DepictsPresenter @Inject constructor(
-            .map { it.distinctBy(DepictedItem::id) }
-    }
 
+        if (media == null) {
```
unit tests?

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -522,6 +522,15 @@ Upload your first media by tapping on the add button.</string>
   <string name="category_edit_helper_edit_message_else">Could not add categories.</string>
   <string name="category_edit_button_text">Update categories</string>
 
+  <string name="depict_edit_helper_make_edit_toast">Trying to update depicts.</string>
```
-> Trying to update depictions

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -522,6 +522,15 @@ Upload your first media by tapping on the add button.</string>
   <string name="category_edit_button_text">Update categories</string>
 
+  <string name="depict_edit_helper_make_edit_toast">Trying to update depicts.</string>
+  <string name="depict_edit_helper_show_edit_title">Depict update</string>
```
-> Edit depictions

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -522,6 +522,15 @@ Upload your first media by tapping on the add button.</string>
+  <string name="depict_edit_helper_make_edit_toast">Trying to update depicts.</string>
+  <string name="depict_edit_helper_show_edit_title">Depict update</string>
+  <plurals name="depict_edit_helper_show_edit_message_if">
+    <item quantity="one">Depict %1$s is added.</item>
```
Depict -> Depiction

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -522,6 +522,15 @@ Upload your first media by tapping on the add button.</string>
+  <string name="depict_edit_helper_show_edit_title">Depict update</string>
+  <plurals name="depict_edit_helper_show_edit_message_if">
+    <item quantity="one">Depict %1$s is added.</item>
+    <item quantity="other">Depicts %1$s are added.</item>
```
Depicts -> Depictions

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -522,6 +522,15 @@ Upload your first media by tapping on the add button.</string>
+    <item quantity="one">Depict %1$s is added.</item>
+    <item quantity="other">Depicts %1$s are added.</item>
+  </plurals>
+  <string name="depict_edit_helper_edit_message_else">Could not add depicts.</string>
```
depicts -> depictions

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -522,6 +522,15 @@ Upload your first media by tapping on the add button.</string>
+    <item quantity="other">Depicts %1$s are added.</item>
+  </plurals>
+  <string name="depict_edit_helper_edit_message_else">Could not add depicts.</string>
+  <string name="depict_edit_button_text">Update depicts</string>
```
Update depicts -> Edit depictions
(unless you think that `Update is more appropriate`, please let me know if so)

Ayan-10 replied:
This ID is not needed now so I removed it.

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -677,6 +686,7 @@ Upload your first media by tapping on the add button.</string>
   <string name="contributions_of_user">Contributions of User: %s</string>
   <string name="achievements_of_user">Achievements of User: %s</string>
   <string name="menu_view_user_page">View user page</string>
+  <string name="edit_depicts">Edit Depicts</string>
```
Edit Depicts -> Edit depictions

It is the same as depict_edit_button_text so maybe factorize and use only one string? This will save our translators a lot of effort.

nicolas-raoul commented:
codeconv says that this pull request reduces unit test coverage significantly... would you mind trying to add more unit tests? @madhurgupta10 any recommendations maybe?

madhurgupta10 commented:
> The diff coverage is `40.93%`.

The diff coverage is `40.93%` which means only about 40% of new lines are tested. Looking at the scope of this PR it is highly important to get around 70-80%

@Ayan-10  Try to break your code blocks into smaller blocks to test individual units (TDD)

Ayan-10 commented:
@madhurgupta10 Thanks a lot for the suggestion. But I think I added tests for most of the new lines. I think the problem here is a lot of lines gets auto intended that's why even if I didn't change anything on those lines, those lines are showing as changed lines. I will try to reverse the indentation. Other than that I added tests for all the lines that I really added or edited. @nicolas-raoul I think that will increase the diff coverage.

Thanks.

nicolas-raoul commented:
Thanks for reversing the indentation changes, that makes the pull request much clearer! :-)

The code coverage now looks much better, but still some effort needed before a good coverage rate is reached. Would you mind following Madhur's advice breaking your code blocks into smaller blocks to test individual units? Thanks a lot!

Ayan-10 commented:
@nicolas-raoul Yes, I can surely break the codes into small test individual units. But according to codecov/patch all the large blocks are already covered, among the left ones most of them are individual units. The problem that I found here is most of the uncovered lines need a real device for testing otherwise they will show errors in unit testing for example `getParentFragment` or `getView`. I tried to cover the lines as much as possible even the single line codes. If I try I can increase the coverage a little more but 70-80% apparently seems not possible.

@madhurgupta10 please have a look and if I am wrong somewhere please give your input.

Thanks a lot.

madhurgupta10 commented:
I think the coverage is sufficient for this PR to be merged, you can add instrumentation tests in future if needed.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsContract.java:
```diff
@@ -40,6 +43,18 @@
          */
         void setDepictsList(List<DepictedItem> depictedItemList);
+
+        Context getFragmentContext();
```
Could you please add javadoc for each added method?
Do not hesitate to copy/paste javadoc from the implementation of each method.
Thanks!

nicolas-raoul commented on app/src/main/res/layout/fragment_media_detail.xml:
```diff
@@ -336,6 +336,15 @@
                 </LinearLayout>
 
+                <Button
+                  android:id="@+id/depictEditButton"
```
depictEditButton -> depictionsEditButton

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:
```diff
@@ -125,6 +125,7 @@ public void onCreate(Bundle savedInstanceState) {
         toolbar.setNavigationOnClickListener(view -> {
             onSupportNavigateUp();
         });
+        applicationKvStore.putBoolean("first_edit_depict", true);
```
Please add a comment explaining what it means when this boolean is true and false.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:
```diff
@@ -517,8 +519,26 @@ private void onDeletionPageExists(Boolean deletionPageExists) {
+     * By clicking on the edit depicts button, it will send user to depict fragment
+     */
+    @OnClick(R.id.depictEditButton)
+    public void onDepictEditButtonClicked() {
```
onDepictEditButtonClicked -> onDepictionsEditButtonClicked

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/notification/NotificationHelper.java:
```diff
@@ -27,6 +27,7 @@
     public static final int NOTIFICATION_EDIT_CATEGORY = 2;
     public static final int NOTIFICATION_EDIT_COORDINATES = 3;
     public static final int NOTIFICATION_EDIT_DESCRIPTION = 4;
+    public static final int NOTIFICATION_EDIT_DEPICT = 5;
```
NOTIFICATION_EDIT_DEPICT -> NOTIFICATION_EDIT_DEPICTIONS

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java:
```diff
@@ -275,6 +293,39 @@ public void onDepictItemClicked(DepictedItem depictedItem) {
+     * @param depictIDs IDs of Depiction
+     * @return Flowable<List<DepictedItem>>
+     */
+    public Flowable<List<DepictedItem>> getDepictions(final List<String> depictIDs){
```
depictIDs -> depictionsQIDs

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java:
```diff
@@ -275,6 +293,39 @@ public void onDepictItemClicked(DepictedItem depictedItem) {
+     * @param depictIDs IDs of depiction ex. ["Q11023","Q1356"]
+     * @return string ex. "Q11023|Q1356"
+     */
+    private String joinIDs(final List<String> depictIDs) {
```
joinIDs -> joinQIDs
depictIDs -> depictionsQIDs

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictEditHelper.kt:
```diff
@@ -0,0 +1,121 @@
+    fun makeDepictEdit(
+        context: Context,
+        media: Media,
+        depicts: List<String>
```
depicts -> depictions

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictEditHelper.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    /**
+     * Appends new depicts
```
depicts -> depictions

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictEditHelper.kt:
```diff
@@ -0,0 +1,121 @@
+     * Appends new depicts
+     *
+     * @param media media
+     * @param depicts to be added
```
depicts -> depictions
Please add a data example.

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -523,6 +523,14 @@ Upload your first media by tapping on the add button.</string>
   <string name="category_edit_button_text">Update categories</string>
 
+  <string name="depict_edit_helper_make_edit_toast">Trying to update depictions.</string>
+  <string name="depict_edit_helper_show_edit_title">Edit depictions</string>
```
depict_edit_helper_show_edit_title -> depictions_edit_helper_show_edit_title

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -523,6 +523,14 @@ Upload your first media by tapping on the add button.</string>
   <string name="category_edit_helper_edit_message_else">Could not add categories.</string>
   <string name="category_edit_button_text">Update categories</string>
 
+  <string name="depict_edit_helper_make_edit_toast">Trying to update depictions.</string>
```
depict_edit_helper_make_edit_toast -> depictions_edit_helper_make_edit_toast

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -523,6 +523,14 @@ Upload your first media by tapping on the add button.</string>
 
+  <string name="depict_edit_helper_make_edit_toast">Trying to update depictions.</string>
+  <string name="depict_edit_helper_show_edit_title">Edit depictions</string>
+  <plurals name="depict_edit_helper_show_edit_message_if">
```
depict_edit_helper_show_edit_message_if -> depictions_edit_helper_show_edit_message_if

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -523,6 +523,14 @@ Upload your first media by tapping on the add button.</string>
+    <item quantity="one">Depiction %1$s is added.</item>
+    <item quantity="other">Depictions %1$s are added.</item>
+  </plurals>
+  <string name="depict_edit_helper_edit_message_else">Could not add depictions.</string>
```
depict_edit_helper_edit_message_else -> depictions_edit_helper_edit_message_else

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -680,6 +688,7 @@ Upload your first media by tapping on the add button.</string>
   <string name="contributions_of_user">Contributions of User: %s</string>
   <string name="achievements_of_user">Achievements of User: %s</string>
   <string name="menu_view_user_page">View user page</string>
+  <string name="edit_depicts">Edit depictions</string>
```
edit_depicts -> edit_depictions

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsPresenter.kt:
```diff
@@ -149,6 +181,76 @@ class DepictsPresenter @Inject constructor(
+     * and saves them in local storage
+     */
+    @SuppressLint("CheckResult")
+    override fun updateDepicts(media: Media) {
```
updateDepicts -> updateDepictions

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsPresenter.kt:
```diff
@@ -149,6 +181,76 @@ class DepictsPresenter @Inject constructor(
+                        })
+                        {
+                            Timber.e(
+                                "Failed to update depicts"
```
depicts -> depictions

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsContract.java:
```diff
@@ -40,6 +43,18 @@
+
+        void goBackToPreviousScreen();
+
+        List<String> getExistingDepicts();
```
getExistingDepicts -> getExistingDepictions

nicolas-raoul reviewed (changes requested):
Minor variable name refactoring

nicolas-raoul commented:
Thanks Ayan! That's a great feature to have, it was not easy but thanks for your perseverance!

Ayan-10 commented:
Thanks for your appreciation.
